modules/data-streams/src/main/java/org/elasticsearch/datastreams/lifecycle/downsampling/ReplaceSourceWithDownsampleIndexTask.java

```diff
@@ -0,0 +1,220 @@
+                    + "Nothing to do here.",
+                sourceBackingIndex,
+                downsampleIndex,
+                dataStreamName,
+                downsampleIndex
+            );
+            return state;
+        }
+        IndexMetadata sourceIndexMeta = state.metadata().index(sourceBackingIndex);
+        DataStream dataStream = state.metadata().dataStreams().get(dataStreamName);
+        if (sourceIndexAbstraction == null) {
+            // index was deleted in the meantime, so let's check if we can make sure the downsample index ends up in the
+            // data stream (if not already there)
+            if (dataStream != null
+                && dataStream.getIndices().stream().filter(index -> index.getName().equals(downsampleIndex)).findAny().isEmpty()) {
+                // add downsample index to data stream
+                LOGGER.trace(
+                    "unable find source index [{}] but adding index [{}] to data stream [{}]",
+                    sourceBackingIndex,
+                    downsampleIndex,
+                    dataStreamName
+                );
+                Metadata.Builder newMetaData = Metadata.builder(state.metadata())
+                    .put(dataStream.addBackingIndex(state.metadata(), downsampleIndexMeta.getIndex()));
+                return ClusterState.builder(state).metadata(newMetaData).build();
+            }
+        } else {
+            // the source index exists
+            DataStream sourceParentDataStream = sourceIndexAbstraction.getParentDataStream();
+            if (sourceParentDataStream != null) {
+                assert sourceParentDataStream.getName().equals(dataStreamName)
+                    : "the backing index must be part of the provided data "
+                        + "stream ["
+                        + dataStreamName
+                        + "] but it is instead part of data stream ["
+                        + sourceParentDataStream.getName()
+                        + "]";
+                if (sourceParentDataStream.getWriteIndex().getName().equals(sourceBackingIndex)) {
+                    String errorMessage = String.format(
+                        Locale.ROOT,
+                        "index [%s] is the write index for data stream [%s] and cannot be replaced",
+                        sourceBackingIndex,
+                        sourceParentDataStream.getName()
+                    );
+                    throw new IllegalStateException(errorMessage);
+                }
+                if (sourceIndexMeta != null) {
+                    // both indices exist, let's copy the origination date from the source index to the downsample index
+                    Metadata.Builder newMetaData = Metadata.builder(state.getMetadata());
+                    IndexMetadata updatedDownsampleMetadata = copyDataStreamLifecycleState(
```

Comment:
Random thought, should we capture the info we need to copy to the new downsampled index when the task is created (inside the task itself) so that if the source index is removed we can still copy the correct lifecycle information into the new downsampled index? (It might be too complicated or introduce a race condition on getting the latest info)

Reply:
Mmmmaybe, do you mind if this is a follow-up ? (we'd have to provide the lifecycle custom and generation time at task creation time and it's quite a large change I believe)

Reply:
Yep, totally okay as a follow-up